compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java

```diff
@@ -0,0 +1,2148 @@
+                }
+            }
+            arrayTypeSymbol.type = arrType;
+            resultType = arrayTypeSymbol.type;
+            if (firstDim) {
+                firstDimArrType = arrType;
+                firstDim = false;
+                continue;
+            }
+            symResolver.markParameterizedType(arrType, arrType.eType);
+        }
+
+        td.setBType(resultType);
+        resolvingTypes.push(resultType);
+
+        if (isError) {
+            return symTable.semanticError;
+        }
+
+        firstDimArrType.eType = resolveTypeDesc(symEnv, mod, typeDefinition, depth + 1, td.elemtype);
+        symResolver.markParameterizedType(firstDimArrType, firstDimArrType.eType);
+        resolvingTypes.pop();
+        return resultType;
+    }
+
+    private BType resolveTypeDesc(BLangTupleTypeNode td, SymbolEnv symEnv, Map<String, BLangNode> mod, int depth,
+                                  BLangTypeDefinition typeDefinition) {
+        currentDepth = depth;
+        if (td.getBType() != null) {
+            return td.getBType();
+        }
+
+        BTypeSymbol tupleTypeSymbol = Symbols.createTypeSymbol(SymTag.TUPLE_TYPE, Flags.asMask(EnumSet.of(Flag.PUBLIC)),
+                Names.EMPTY, symEnv.enclPkg.symbol.pkgID, null,
+                symEnv.scope.owner, td.pos, BUILTIN);
+        List<BTupleMember> memberTypes = new ArrayList<>();
+        BTupleType tupleType = new BTupleType(tupleTypeSymbol, memberTypes);
+        tupleTypeSymbol.type = tupleType;
+        td.setBType(tupleType);
+        resolvingTypes.push(tupleType);
+
+        for (BLangSimpleVariable memberNode: td.getMemberNodes()) {
+            BType type = resolveTypeDesc(symEnv, mod, typeDefinition, depth + 1, memberNode.typeNode);
+            SymbolEnv tupleEnv = SymbolEnv.createTypeEnv(td, new Scope(tupleTypeSymbol), symEnv);
+            symEnter.defineNode(memberNode, tupleEnv);
+            BVarSymbol varSymbol = new BVarSymbol(memberNode.getBType().flags, memberNode.symbol.name,
+                    memberNode.symbol.pkgID, memberNode.getBType(), memberNode.symbol.owner, memberNode.pos, SOURCE);
+            memberTypes.add(new BTupleMember(type, varSymbol));
+        }
+
+        if (td.restParamType != null) {
+            BType tupleRestType = resolveTypeDesc(symEnv, mod, typeDefinition, depth + 1, td.restParamType);
+            if (tupleRestType.tag == TypeTags.SEMANTIC_ERROR) {
+                return symTable.semanticError;
+            }
+            tupleType.restType = tupleRestType;
+            symResolver.markParameterizedType(tupleType, tupleType.restType);
+        }
+
+        symResolver.markParameterizedType(tupleType, tupleType.getTupleTypes());
+        resolvingTypes.pop();
+        return tupleType;
+    }
+
+    private BType resolveTypeDesc(BLangRecordTypeNode td, SymbolEnv symEnv, Map<String, BLangNode> mod, int depth,
+                                  BLangTypeDefinition typeDefinition) {
```

Comment:
Shall we create a class to encapsulate `symEnv, mod, depth, typeDefinition` like we created the `AnalyzerData` class?

Reply:
Updated 70dc2c1266fafd2aedf3400af1f3bc6d778de232